Component graph implementation

shsms opened the pull request:
This implementation has a similar interface to the one in the SDK.  It has validation and supports traversal.

shsms committed:
Introduce `Node` and `Edge` traits

The `ComponentGraph` implementation will operate on types that
implement these traits.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Introduce an `Error` type

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Add component category predicates on `Node` types

These make category checks easier when traversing the graph.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Introduce the `ComponentGraph` struct

This commit also adds the `try_new` method for creating
`ComponentGraph` instances.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Add methods for retrieving components and connections

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Validate acyclicity and connectedness of `ComponentGraph`s

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Add generic invariant checks for the relationship between components

These will be used in validating that components have the right kind
of neighbors.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Validate components have sensible neighbors

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Add methods for checking the roles of meters in the graph

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Add module docs

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Add a CI file

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms commented:
This can completely replace the component graph in the SDK.  Formula building will come subsequently.

tiyash-basu-frequenz commented on .github/workflows/ci.yaml:
```diff
@@ -0,0 +1,25 @@
+        run: cargo clippy
+
+      - name: Run tests
+        run: cargo test
```
Would suggest adding a step to check formatting.

llucax replied:
Maybe we should open source our actions and/or reusable workflows for rust.

tiyash-basu-frequenz replied:
Agreed! I already plan to start with it next week.

shsms replied:
done

tiyash-basu-frequenz commented on src/graph/creation.rs:
```diff
@@ -34,6 +34,8 @@ where
         };
         cg.add_connections(connections)?;
 
+        validate(&cg)?;
```
Why not implement `validate()` as a member function of `ComponentGraph`?

shsms replied:
done, I think it was because I thought there would be too many validation methods.  But you're right, as long as the individual validation methods are elsewhere, this one can be under `ComponentGraph`.

llucax commented on .github/workflows/ci.yaml:
```diff
@@ -0,0 +1,25 @@
+
+jobs:
+  test:
+    runs-on: ubuntu-latest
```
I would be on the conservative side and pin this to some concrete ubuntu version.

shsms replied:
Pinned to 20.04.  Still have 6 months before that reaches end of life. \o/

llucax replied:
Why not 22.04 or even 24.04 then?

llucax commented on Cargo.toml:
```diff
@@ -0,0 +1,11 @@
+[package]
+name = "frequenz-microgrid-component-graph-rs"
```
Do we want to keep the `-rs` suffix in the package name? At least in Python we remove the suffix.

shsms replied:
done.

llucax commented on Cargo.toml:
```diff
@@ -0,0 +1,11 @@
+edition = "2021"
+
+[lib]
+name = "component_graph"
```
Is this name top-level or namespaced? If it is not namespaced, wouldn't it make sense to put it inside a `frequenz` namespace like we do in Python? Also mirroring the package name, maybe even `frequenz/microgrid` / `frequenz::microgrid`?

shsms replied:
We can't have different libraries in Rust under the same namespace (module).  I can change the library name to `frequenz_component_graph`, but then we'd have to:

```rust
use frequenz_component_graph;

let graph = frequenz_component_graph::ComponentGraph(...);
```

or 

```rust
use frequenz_component_graph as cg;
```

It depends on how strongly you feel about this.

llucax replied:
I see. IMHO `component_graph` with nothing else sounds a bit too generic. I'm not sure what is the common approach to this for companies that maybe have many open source rust libraries. From the code perspective, if it is easy enough to create a short alias as you showed, I guess we can live with the long name. Otherwise I noticed in some places we seem to use at least `fz_` as a prefix/namespace for our stuff, maybe `fz_component_graph` would be a OK-ish balance if having long library names is not popular at all in rust?

I just did a quick search in crates.io for google stuff, and even when it doesn't seem to be official, most libraries seem to use a full `google_cloud_xxx` or `googleapis_tonic_google_firestore_v1` as module, so I guess `frequenz_microgrid_component_graph` doesn't look that alien to rust. I would go with that to make it consistent with what we do in Python, but I could also live with `fz_component_graph` if it were necessary.

shsms replied:
done

llucax commented on src/component_category.rs:
```diff
@@ -0,0 +1,121 @@
+    Unspecified,
+    Grid,
+    Meter,
+    Battery,
```
We also have `BatteryType` (in the new microgrid API version), I would do the same for this type, even when in principle it doesn't seem to be as useful as with inverters, for the sake of consistency and in case it becomes more important in the future.

shsms replied:
done.

llucax commented on src/component_category.rs:
```diff
@@ -0,0 +1,121 @@
+    Meter,
+    Battery,
+    Inverter(InverterType),
+    EvCharger,
```
We also have different `EvChargerType`s.

shsms replied:
done.

nhatcher-frequenz commented on .github/workflows/ci.yaml:
```diff
@@ -0,0 +1,25 @@
+          submodules: true
+
+      - name: Run linter
+        run: cargo clippy
```
I like being a bit more restrictive here:

```
cargo clippy -- -W clippy::unwrap_used -W clippy::expect_used -W clippy::panic
```

Or some variation of it. A fairly restrictive option that gives very good tips soemtimes:

```
cargo clippy -- -W clippy::restriction
```

shsms replied:
Wow, didn't know about those.  I've added a fixup commit, will squash later.

llucax commented on src/graph/creation.rs:
```diff
@@ -0,0 +1,254 @@
+        for component in components {
+            let cid = component.component_id();
+
+            if component.is_unspecified() {
+                return Err(Error::invalid_component(format!(
+                    "ComponentCategory not specified for component: {cid}"
+                )));
```
I wonder if having a unspecified component (or inverter) is an error fatal enough to completely refuse to create the graph. Couldn't we just emit an error and continue and somehow deal with unspecified components later, depending on the operation we want to do? Specially if a unspecified component is a leaf it looks like it might be harmless to ignore it (maybe we'll get wrong consumption or production, but better than nothing, right?)

shsms replied:
I agree, will make it configurable in a subsequent PR.

tiyash-basu-frequenz commented on src/graph/meter_roles.rs:
```diff
@@ -0,0 +1,434 @@
+    /// A meter is identified as a grid meter if:
+    ///   - it is a successor of the grid component,
+    ///   - all its siblings are meters,
+    ///   - if there are siblings, the successors of it and the successors of
+    ///     its siblings are meters.
```
According to this specification, there can be multiple "grid meters", and the grid connection's total load would be the sum of all these meters.

This terminology seems a bit odd because if we call a meter a "grid meter", then we should only have to read this one meter to get the net grid activity.

Can we improve it?

shsms replied:
Wait, I thought we had cases like this, with multiple grid meters.  And you mentioned the "diamond" configuration for grid meters as well, right?  How do you suggest we change it?

But anyway, this particular method is there just so the existing SDK formulas can continue to work. It will go away as soon as the new approach for building formulas is implemented.  So we don't have to put in too much energy into this one.

llucax commented on src/graph/validation.rs:
```diff
@@ -0,0 +1,46 @@
+
+    let validator = ComponentGraphValidator { cg, root };
+
+    validator.validate_acyclicity(root, vec![])?;
+    validator.validate_connected_graph(root)?;
+
+    validator.validate_root()?;
+    validator.validate_meters()?;
+    validator.validate_inverters()?;
+    validator.validate_batteries()?;
+    validator.validate_ev_chargers()?;
+    validator.validate_chps()?;
```
Not sure how this is normally done in rust, or if there is a popular validation library to use, but when validating, in particular if at some point the validation result needs to be shown to the users, it might be nice to be able to collect all validation errors and show all instead of bailing out on the first error. Maybe for *acyclicity* and *connectivity* it makes sense to bail out because all the other validation assumes a a DAG though.

tiyash-basu-frequenz replied:
Makes sense. Then you get all the validation issues in one go, instead of finding them out one by one.

llucax replied:
This is the only pending comment from me but I think it can be perfectly left for a future improvement. Let me know if I should create an issue about it. So LGTM, will leave the approval to a rusty person.

shsms replied:
Yes, I think this fits well with the config support I'm already working on, so I will do it there.

tiyash-basu-frequenz commented on src/graph/meter_roles.rs:
```diff
@@ -0,0 +1,434 @@
+        (components, connections)
+    }
+
+    fn assert_meter_role(
+        components: Vec<TestComponent>,
+        connections: Vec<TestConnection>,
+        filter: impl Fn(&ComponentGraph<TestComponent, TestConnection>, u64) -> Result<bool, Error>,
+        expected_grid_meters: Vec<u64>,
+    ) -> Result<(), Error> {
+        let graph = ComponentGraph::try_new(components.clone(), connections.clone())?;
+
+        let mut grid_meters = vec![];
+        for comp in graph.components() {
+            if filter(&graph, comp.component_id())? {
+                grid_meters.push(comp.component_id());
+            }
+        }
+        assert_eq!(grid_meters, expected_grid_meters);
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_is_grid_meter() -> Result<(), Error> {
+        let (components, connections) = nodes_and_edges();
+        assert_meter_role(
+            components,
+            connections,
+            ComponentGraph::is_grid_meter,
+            vec![2],
+        )?;
```
If the assertion in `assert_meter_role` fails, the error will show line 285, and not thisline, This would make the error difficult to identify.

I would recommend returning a result an error in line 285 instead.

shsms replied:
true, done.

llucax commented on src/graph/validation/invariant_checks.rs:
```diff
@@ -0,0 +1,148 @@
+    }
+
+    /// Checks that the given node only has predecessors with the given categories.
+    pub(super) fn ensure_predecessor_categories(
```
Wouldn't it make sense to generalize this to ensure a predecessor (or successor) fulfills some predicate (function)?

shsms replied:
that makes sense, especially when we want to introduce BatteryType, etc.

done.

llucax commented on src/graph/validation/validate_graph.rs:
```diff
@@ -0,0 +1,243 @@
+        node: &N,
+        mut predecessors: Vec<u64>,
+    ) -> Result<(), Error> {
+        predecessors.push(node.component_id());
```
Doesn't petgraph provide a check for this already? https://docs.rs/petgraph/latest/petgraph/algo/fn.is_cyclic_directed.html

If it is not suitable for us for some reason maybe add a comment to clarify as maybe other people try to use it?

shsms replied:
Their implementation doesn't show where the cycle is, this implementation gives us better logs.

llucax commented on src/graph/validation/validate_neighbors.rs:
```diff
@@ -0,0 +1,404 @@
+        Ok(())
+    }
+
+    pub(super) fn validate_meters(&self) -> Result<(), Error> {
```
Maybe document what kind validations are done here like with `is_xxx_meter()`? (same for other `validate_xxx()`.

shsms replied:
done

llucax commented on src/graph/validation/validate_neighbors.rs:
```diff
@@ -0,0 +1,404 @@
+            )?;
+
+            match inverter_type {
+                InverterType::Battery => {
+                    self.ensure_not_leaf(inverter)?;
+                    self.ensure_successor_categories(inverter, &[ComponentCategory::Battery])?;
```
Also in terms of resilience, is there anything preventing us to just emit an error and continue and assume the inverter is a "<whatever the successor is> inverter"? Most likely it was an error when building the component graph, it would be like duck-typing, if we use it as a "<whatever the successor is> inverter" and quacks as a "<whatever the successor is> inverter", why not continuing? If it doesn't quack as we want it to quack, we'll emit some other errors when using it and getting errors.

shsms replied:
My plan was to introduce a `Config` param and a new `ComponentGraph::try_new_with_config()` constructor, that would allow configuring various behaviours, like `validation.allow_disconnected_components`, `unspecified_inverters_are_battery_inverters`, `validation.warn_and_continue`, etc.  Then we can configure them per-site if required.

llucax reviewed (commented):
I did a shallow review, not getting much into the rust code itself.

Also many of the comments/questions are definitely not to address in this PR but more as a brain storming and possibly create separate issues to discuss them. If you foresee an upcoming lengthy discussion maybe just create a different issue first and defer the discussion to the issue directly.

shsms committed:
Make `validate` a method of `ComponentGraph`

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Add support for `BatteryType`

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Add support for `EvChargerType`

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Move types used in tests to a shared `test_types` module

Just like the unit tests modules, the `test_types` module is also
conditionally compiled, only when running the tests.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Update readme

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

tiyash-basu-frequenz reviewed (approved):
LGTM!
